File: src/main/java/ro/nicolaemariusghergu/queryit/service/impl/ManufacturerServiceImpl.java
package ro.nicolaemariusghergu.queryit.service.impl;

import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import ro.nicolaemariusghergu.queryit.dto.ManufacturerDto;
import ro.nicolaemariusghergu.queryit.mapper.ManufacturerMapper;
import ro.nicolaemariusghergu.queryit.repository.ManufacturerRepository;
import ro.nicolaemariusghergu.queryit.service.ManufacturerService;

import java.util.List;
import java.util.NoSuchElementException;

@Service
public record ManufacturerServiceImpl(ManufacturerRepository manufacturerRepository) implements ManufacturerService {

    @Override
    public ResponseEntity<ManufacturerDto> findManufacturerById(Long id) {
        return ResponseEntity.ok(manufacturerRepository.findById(id).stream()
                .map(ManufacturerMapper.INSTANCE::manufacturerToManufacturerDto)
                .findFirst()
                .orElseThrow(
                        () -> new NoSuchElementException("Manufacturer does not exist!"))
        );
    }

    @Override
    public ResponseEntity<List<ManufacturerDto>> getManufacturers() {
        return ResponseEntity.ok(manufacturerRepository.findAll()
                .stream()
                .map(ManufacturerMapper.INSTANCE::manufacturerToManufacturerDto)
                .toList());
    }

    @Override
    public ResponseEntity<Long> addManufacturer(ManufacturerDto manufacturerDto) {
        manufacturerRepository.save(ManufacturerMapper.INSTANCE.manufacturerDtoToManufacturer(manufacturerDto));
        return ResponseEntity.ok(manufacturerDto.getId());
    }

    @Override
    public ResponseEntity<ManufacturerDto> getManufacturerByName(String name) {
        return ResponseEntity.ok(ManufacturerMapper.INSTANCE.manufacturerToManufacturerDto(
                manufacturerRepository.findByName(name)
                        .orElseThrow(() ->
                                new NoSuchElementException("Manufacturer cannot be found"))
        ));
    }

    @Override
    public ResponseEntity<Long> deleteManufacturerById(Long id) {
        manufacturerRepository.deleteById(id);
        return ResponseEntity.ok(id);
    }

    @Override
    public ResponseEntity<ManufacturerDto> updateManufacturer(ManufacturerDto manufacturerDto) {
        ManufacturerDto modifiedManufacturer = manufacturerRepository.findById(manufacturerDto.getId()).stream()
                .map(ManufacturerMapper.INSTANCE::manufacturerToManufacturerDto)
                .map(updatedManufacturer -> {
                    updatedManufacturer.setName(manufacturerDto.getName());
                    return updatedManufacturer;
                })
                .findAny()
                .orElseThrow(() ->
                        new NoSuchElementException("Manufacturer has not been found")
                );
        manufacturerRepository.save(
                ManufacturerMapper.INSTANCE.manufacturerDtoToManufacturer(modifiedManufacturer));
        return ResponseEntity.ok(modifiedManufacturer);
    }
}

File: src/main/kotlin/ro/nicolaemariusghergu/queryit/service/impl/ManufacturerServiceImpl.kt
package ro.nicolaemariusghergu.queryit.service.impl

import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import ro.nicolaemariusghergu.queryit.dto.ManufacturerDto
import ro.nicolaemariusghergu.queryit.mapper.ManufacturerMapper
import ro.nicolaemariusghergu.queryit.model.Manufacturer
import ro.nicolaemariusghergu.queryit.repository.ManufacturerRepository
import ro.nicolaemariusghergu.queryit.service.ManufacturerService

@Service
class ManufacturerServiceImpl(private val manufacturerRepository: ManufacturerRepository) : ManufacturerService {
    override fun findManufacturerById(id: Long): ResponseEntity<ManufacturerDto> {
        return ResponseEntity.ok(manufacturerRepository.findById(id).stream()
                .map { manufacturer: Manufacturer -> ManufacturerMapper.INSTANCE.manufacturerToManufacturerDto(manufacturer) }
                .findFirst()
                .orElseThrow { NoSuchElementException("Manufacturer does not exist!") }
        )
    }

    override fun getManufacturers(): ResponseEntity<List<ManufacturerDto>> {
        return ResponseEntity.ok(manufacturerRepository.findAll()
                .stream()
                .map { manufacturer: Manufacturer -> ManufacturerMapper.INSTANCE.manufacturerToManufacturerDto(manufacturer) }
                .toList())
    }

    override fun addManufacturer(manufacturerDto: ManufacturerDto): ResponseEntity<Long> {
        manufacturerRepository.save(ManufacturerMapper.INSTANCE.manufacturerDtoToManufacturer(manufacturerDto))
        return ResponseEntity.ok(manufacturerDto.id)
    }

    override fun getManufacturerByName(name: String): ResponseEntity<ManufacturerDto> {
        return ResponseEntity.ok(ManufacturerMapper.INSTANCE.manufacturerToManufacturerDto(
                manufacturerRepository.findByName(name)
                        .orElseThrow { NoSuchElementException("Manufacturer cannot be found") }
        ))
    }

    override fun deleteManufacturerById(id: Long): ResponseEntity<Long> {
        manufacturerRepository.deleteById(id)
        return ResponseEntity.ok(id)
    }

    override fun updateManufacturer(manufacturerDto: ManufacturerDto): ResponseEntity<ManufacturerDto> {
        val modifiedManufacturer: Manufacturer = manufacturerRepository.findById(manufacturerDto.id).stream()
                .map { manufacturer: Manufacturer ->
                    manufacturer.name = manufacturerDto.name
                    manufacturer
                }
                .findAny()
                .orElseThrow { NoSuchElementException("Manufacturer has not been found") }
        manufacturerRepository.save(modifiedManufacturer)
        return ResponseEntity.ok(ManufacturerMapper.INSTANCE.manufacturerToManufacturerDto(modifiedManufacturer))
    }
}